### PR TITLE
Add restricted tanks, put them into use with the Lava Coil and enhanced ...

### PR DESCRIPTION
**Pull request by iTitus**
...tank widgets tooltips.

### CODE REVIEW

**Comment by BigXplosion** [comment]
had a quick look and seemed good to me.

File: src/main/java/buildcraftAdditions/tileEntities/TileLavaCoil.java [hunk 1]
```diff
@@ -68,50 +78,30 @@ public void updateEntity() {
...
-				return amount;
-			}
-		}
-		return 0;
+		return tank.fill(resource, doFill);
 	}
 
 	@Override
 	public FluidStack drain(ForgeDirection from, FluidStack resource, boolean doDrain) {
+		if (resource != null) {
```
Comment (AEnterprise) [inline src/main/java/buildcraftAdditions/tileEntities/TileLavaCoil.java]: wouldn't it be easier if you do this in the tank? also you can just return null here as nobody uses this as far as i know

File: src/main/java/buildcraftAdditions/utils/RestrictedTank.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	public boolean canFill(Fluid fluid) {
+		return acceptor.accepts(new FluidStack(fluid.getID(), 1));
+	}
+
+	@Override
+	public String toString() {
+		return super.toString() + ", Restrictions: [" + acceptor.getDescription() + "]";
+	}
+
+	public static interface FluidAcceptor {
```
Comment (AEnterprise) [inline src/main/java/buildcraftAdditions/utils/RestrictedTank.java]: i prefer that interfaces are not in the class files, makes them easier to find later

File: src/main/java/buildcraftAdditions/utils/BlacklistedTank.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import net.minecraftforge.fluids.FluidStack;
+
+/**
+ * Copyright (c) 2014-2015, AEnterprise
+ * http://buildcraftadditions.wordpress.com/
+ * Buildcraft Additions is distributed under the terms of GNU GPL v3.0
+ * Please check the contents of the license located in
+ * http://buildcraftadditions.wordpress.com/wiki/licensing-stuff/
+ */
+public class BlacklistedTank extends RestrictedTank {
```
Comment (AEnterprise) [inline src/main/java/buildcraftAdditions/utils/BlacklistedTank.java]: blacklisted tank and restricted tank? why not make it all one class? i have no idea why there would ever be another type of restricted tank, also the name is confusing it should be a whitelisted tank

**Comment by AEnterprise** [comment]
added some line notes, i can see the use for a restricted tank thing but you overcomplicated it, the interface isn't realy needed as it only has one use that that is for the restricted tank, the blacklist tank should be a whitelisted tank as it only accept certain fluids, a blacklist tank should accept all fluids except the fluids that you tell it to accept

are you gona change that or should i implement it myself? the idea itself is good though

**Comment by iTitus** [comment]
It is more flexible

**Comment by iTitus** [comment]
The restricted tank is a superclass which contains all the logic. The sub-classes (only made examples) only need to give it a FluidAcceptor (could also be an abstract class). You then can create new FluidAcceptors for the coolants or the recipes...
As for the Dusts there are easier ways to do this, but this way gives maximum amount of flexibility.

File: src/main/java/buildcraftAdditions/utils/BlacklistedTank.java (hunk 3 above)
Comment (iTitus) [inline src/main/java/buildcraftAdditions/utils/BlacklistedTank.java]: BlacklistedTank: All fluids given as an argument CANNOT go into it. All others can.
WhitelistedTank: The fluids given as an argument are the ONLY fluid(s) that are allowed in the tank.

File: src/main/java/buildcraftAdditions/utils/RestrictedTank.java (hunk 2 above)
Comment (iTitus) [inline src/main/java/buildcraftAdditions/utils/RestrictedTank.java]: Will do

File: src/main/java/buildcraftAdditions/tileEntities/TileLavaCoil.java (hunk 1 above)
Comment (iTitus) [inline src/main/java/buildcraftAdditions/tileEntities/TileLavaCoil.java]: Will do

File: src/main/java/buildcraftAdditions/utils/RestrictedTank.java (hunk 2 above)
Comment (AEnterprise) [inline src/main/java/buildcraftAdditions/utils/RestrictedTank.java]: alos it's not realy a needed interface

File: src/main/java/buildcraftAdditions/utils/BlacklistedTank.java (hunk 3 above)
Comment (AEnterprise) [inline src/main/java/buildcraftAdditions/utils/BlacklistedTank.java]: indeed that's what i want, you are using this tank in the lavaCoil so only lava should go in there

**Comment by AEnterprise** [comment]
i like some flexibility but not over-flexible that makes things more complicated then needed

**Comment by iTitus** [comment]
Its very easy

**Comment by iTitus** [comment]
Should I add th restricted tanks to the refineries?

File: src/main/java/buildcraftAdditions/utils/RestrictedTank.java (hunk 2 above)
Comment (BigXplosion) [inline src/main/java/buildcraftAdditions/utils/RestrictedTank.java]: why are you so agains classes and interfaces in other class files?

**Comment by BigXplosion** [comment]
@iTitus, I think you can that would make this PR complete

**Comment by AEnterprise** [comment]
know what? i'll just pull this down, merge it and finish it myself quickly, might rewrite some of this in the process

**Comment by BigXplosion** [comment]
make sure @iTitus knows

**Comment by AEnterprise** [comment]
he made the pr so he should get the notifications

**Pull request closed without merging**